File: packages/go/protocol/protocol.go
package protocol

import (
	"encoding/json"
	"fmt"
)

const Version = 1
const MaxFrameSize = 32 * 1024 * 1024

type MessageType string

const (
	MessageTypeAuth                MessageType = "auth"
	MessageTypeAuthResponse        MessageType = "auth_response"
	MessageTypeVersionNegotiate    MessageType = "version_negotiate"
	MessageTypeVersionNegotiateAck MessageType = "version_negotiate_ack"
	MessageTypeFlowControl         MessageType = "flow_control"
	MessageTypeOpenTunnel          MessageType = "open_tunnel"
	MessageTypeOpenTunnelAck       MessageType = "open_tunnel_ack"
	MessageTypeCloseTunnel         MessageType = "close_tunnel"
	MessageTypeData                MessageType = "data"
	MessageTypeHeartbeat           MessageType = "heartbeat"
	MessageTypeError               MessageType = "error"
	MessageTypeHTTPRequest         MessageType = "http_request"
	MessageTypeHTTPResponse        MessageType = "http_response"
	MessageTypeTCPData             MessageType = "tcp_data"
	MessageTypeTCPClose            MessageType = "tcp_close"
	MessageTypeUDPData             MessageType = "udp_data"
	MessageTypeUDPResponse         MessageType = "udp_response"
)

var messageTypes = map[MessageType]struct{}{
	MessageTypeAuth: {}, MessageTypeAuthResponse: {},
	MessageTypeVersionNegotiate: {}, MessageTypeVersionNegotiateAck: {},
	MessageTypeFlowControl: {}, MessageTypeOpenTunnel: {},
	MessageTypeOpenTunnelAck: {}, MessageTypeCloseTunnel: {},
	MessageTypeData: {}, MessageTypeHeartbeat: {}, MessageTypeError: {},
	MessageTypeHTTPRequest: {}, MessageTypeHTTPResponse: {},
	MessageTypeTCPData: {}, MessageTypeTCPClose: {},
	MessageTypeUDPData: {}, MessageTypeUDPResponse: {},
}

type Envelope struct {
	Version   int             `json:"version"`
	Type      MessageType     `json:"type"`
	RequestID string          `json:"request_id,omitempty"`
	Payload   json.RawMessage `json:"payload,omitempty"`
}

type AuthRequest struct {
	Token                 string   `json:"token"`
	AgentID               string   `json:"agent_id,omitempty"`
	RequestedCapabilities []string `json:"requested_capabilities,omitempty"`
}
type AuthResponse struct {
	Authenticated       bool     `json:"authenticated"`
	AgentID             string   `json:"agent_id,omitempty"`
	OrganizationID      string   `json:"organization_id,omitempty"`
	GrantedCapabilities []string `json:"granted_capabilities,omitempty"`
	Error               string   `json:"error,omitempty"`
}
type VersionRequest struct {
	MinVersion    int    `json:"min_version"`
	MaxVersion    int    `json:"max_version"`
	ClientName    string `json:"client_name,omitempty"`
	ClientVersion string `json:"client_version,omitempty"`
}
type VersionAck struct {
	NegotiatedVersion int   `json:"negotiated_version"`
	SupportedVersions []int `json:"supported_versions"`
}
type OpenRequest struct {
	Token        string `json:"token"`
	TunnelID     string `json:"tunnel_id,omitempty"`
	LocalPort    int    `json:"local_port"`
	Protocol     string `json:"protocol"`
	Subdomain    string `json:"subdomain,omitempty"`
	CustomDomain string `json:"custom_domain,omitempty"`
	Password     string `json:"password,omitempty"`
}
type OpenAck struct {
	TunnelID   string `json:"tunnel_id"`
	PublicURL  string `json:"public_url"`
	PublicPort int    `json:"public_port,omitempty"`
}
type CloseRequest struct {
	TunnelID string `json:"tunnel_id"`
	Reason   string `json:"reason,omitempty"`
}
type HeartbeatRequest struct {
	Timestamp int64 `json:"timestamp"`
}
type ErrorMessage struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

type FlowControl struct {
	StreamID   string `json:"stream_id"`
	Action     string `json:"action"`
	WindowSize int64  `json:"window_size,omitempty"`
}
type Data struct {
	TunnelID string `json:"tunnel_id"`
	StreamID string `json:"stream_id"`
	Data     string `json:"data"`
}
type HTTPRequest struct {
	Method  string              `json:"method"`
	Path    string              `json:"path"`
	Headers map[string][]string `json:"headers"`
	Body    string              `json:"body,omitempty"`
}
type HTTPResponse struct {
	StatusCode int                 `json:"status_code"`
	Headers    map[string][]string `json:"headers"`
	Body       string              `json:"body,omitempty"`
	Error      string              `json:"error,omitempty"`
}
type TCPData struct {
	TunnelID     string `json:"tunnel_id,omitempty"`
	ConnectionID string `json:"connection_id"`
	Data         string `json:"data"`
}
type TCPClose struct {
	TunnelID     string `json:"tunnel_id,omitempty"`
	ConnectionID string `json:"connection_id"`
	Reason       string `json:"reason,omitempty"`
}
type UDPData struct {
	TunnelID      string `json:"tunnel_id,omitempty"`
	PacketID      string `json:"packet_id"`
	SourceAddress string `json:"source_address"`
	SourcePort    int    `json:"source_port"`
	Data          string `json:"data"`
}
type UDPResponse struct {
	TunnelID      string `json:"tunnel_id,omitempty"`
	PacketID      string `json:"packet_id"`
	TargetAddress string `json:"target_address"`
	TargetPort    int    `json:"target_port"`
	Data          string `json:"data"`
}

func Encode(message Envelope) ([]byte, error) {
	if message.Version == 0 {
		message.Version = Version
	}
	if message.Type == "" {
		return nil, fmt.Errorf("protocol message type is required")
	}
	encoded, err := json.Marshal(message)
	if err != nil {
		return nil, fmt.Errorf("encode protocol message: %w", err)
	}
	if len(encoded) > MaxFrameSize {
		return nil, fmt.Errorf("protocol frame exceeds %d bytes", MaxFrameSize)
	}
	return encoded, nil
}

func Decode(data []byte) (Envelope, error) {
	if len(data) > MaxFrameSize {
		return Envelope{}, fmt.Errorf("protocol frame exceeds %d bytes", MaxFrameSize)
	}
	var message Envelope
	if err := json.Unmarshal(data, &message); err != nil {
		return Envelope{}, fmt.Errorf("decode protocol message: %w", err)
	}
	if message.Version != Version || message.Type == "" {
		return Envelope{}, fmt.Errorf("invalid protocol envelope")
	}
	if _, ok := messageTypes[message.Type]; !ok {
		return Envelope{}, fmt.Errorf("invalid protocol message type %q", message.Type)
	}
	return message, nil
}

func EncodePayload(messageType MessageType, requestID string, payload any) ([]byte, error) {
	body, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("encode protocol payload: %w", err)
	}
	return Encode(Envelope{Type: messageType, RequestID: requestID, Payload: body})
}

func DecodePayload(message Envelope, target any) error {
	if err := json.Unmarshal(message.Payload, target); err != nil {
		return fmt.Errorf("decode protocol payload: %w", err)
	}
	return nil
}

File: packages/rust/src/protocol.rs
use serde::{de::DeserializeOwned, Deserialize, Serialize};

pub const VERSION: u8 = 1;
pub const MAX_FRAME_SIZE: usize = 32 * 1024 * 1024;

const MESSAGE_TYPES: &[&str] = &[
    "auth",
    "auth_response",
    "version_negotiate",
    "version_negotiate_ack",
    "flow_control",
    "open_tunnel",
    "open_tunnel_ack",
    "close_tunnel",
    "data",
    "heartbeat",
    "error",
    "http_request",
    "http_response",
    "tcp_data",
    "tcp_close",
    "udp_data",
    "udp_response",
];

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Envelope<T = serde_json::Value> {
    pub version: u8,
    #[serde(rename = "type")]
    pub message_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub request_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub payload: Option<T>,
}

pub fn encode<T: Serialize>(message: &Envelope<T>) -> crate::Result<Vec<u8>> {
    if message.version != VERSION {
        return Err(crate::Error::Protocol(
            "unsupported protocol version".into(),
        ));
    }
    if !MESSAGE_TYPES.contains(&message.message_type.as_str()) {
        return Err(crate::Error::Protocol("message type is required".into()));
    }
    let data = serde_json::to_vec(message)?;
    if data.len() > MAX_FRAME_SIZE {
        return Err(crate::Error::Protocol("frame exceeds maximum size".into()));
    }
    Ok(data)
}

pub fn decode<T: DeserializeOwned>(data: &[u8]) -> crate::Result<Envelope<T>> {
    if data.len() > MAX_FRAME_SIZE {
        return Err(crate::Error::Protocol("frame exceeds maximum size".into()));
    }
    let message: Envelope<T> = serde_json::from_slice(data)?;
    if message.version != VERSION || !MESSAGE_TYPES.contains(&message.message_type.as_str()) {
        return Err(crate::Error::Protocol("invalid protocol envelope".into()));
    }
    Ok(message)
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct AuthRequest {
    pub token: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub agent_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub requested_capabilities: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct VersionNegotiate {
    pub min_version: u8,
    pub max_version: u8,
    pub client_name: String,
    pub client_version: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct AuthResponse {
    pub authenticated: bool,
    pub agent_id: Option<String>,
    pub organization_id: Option<String>,
    pub granted_capabilities: Option<Vec<String>>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct VersionNegotiateAck {
    pub negotiated_version: u8,
    pub supported_versions: Vec<u8>,
    pub server_name: Option<String>,
    pub server_version: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct OpenTunnel {
    pub token: String,
    pub local_port: u16,
    pub protocol: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tunnel_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub subdomain: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub custom_domain: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub password: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct OpenTunnelAck {
    pub tunnel_id: String,
    pub public_url: String,
    pub public_port: Option<u16>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct FlowControl {
    pub stream_id: String,
    pub action: String,
    pub window_size: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Data {
    pub tunnel_id: String,
    pub stream_id: String,
    pub data: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct HttpRequest {
    pub method: String,
    pub path: String,
    pub headers: std::collections::HashMap<String, Vec<String>>,
    pub body: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct HttpResponse {
    pub status_code: u16,
    pub headers: std::collections::HashMap<String, Vec<String>>,
    pub body: Option<String>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TcpData {
    pub tunnel_id: Option<String>,
    pub connection_id: String,
    pub data: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TcpClose {
    pub tunnel_id: Option<String>,
    pub connection_id: String,
    pub reason: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct UdpData {
    pub tunnel_id: Option<String>,
    pub packet_id: String,
    pub source_address: String,
    pub source_port: u16,
    pub data: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct UdpResponse {
    pub tunnel_id: Option<String>,
    pub packet_id: String,
    pub target_address: String,
    pub target_port: u16,
    pub data: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CloseTunnel {
    pub tunnel_id: String,
    pub reason: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Heartbeat {
    pub timestamp: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ErrorMessage {
    pub code: String,
    pub message: String,
}
